File: medium/1833-maximum-ice-cream-bars/solution.py
from typing import List


class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        """
        1833. Maximum Ice Cream Bars
        Time: O(n + m)
        Space: O(m)
        """
        max_cost = max(costs)
        frequency = [0] * (max_cost + 1)
        for cost in costs:
            frequency[cost] += 1

        answer = 0

        for cost in range(1, max_cost + 1):
            total_cost = cost * frequency[cost]
            if coins >= total_cost:
                coins -= total_cost
                answer += frequency[cost]
            else:
                answer += coins // cost
                break

        return answer

File: medium/1833-maximum-ice-cream-bars/solution.java
/**
 * 1833. Maximum Ice Cream Bars
 * Time: O(n + m)
 * Space: O(m)
 */
class Solution {
    public int maxIceCream(int[] costs, int coins) {
        int maxCost = 0;
        for (int c : costs) {
            maxCost = Math.max(maxCost, c);
        }

        int[] freq = new int[maxCost + 1];
        for (int c : costs) {
            freq[c]++;
        }

        int count = 0;

        for (int cost = 1; cost <= maxCost; cost++) {
            if (freq[cost] == 0) continue;

            long totalCost = (long) cost * freq[cost];

            if (coins >= totalCost) {
                coins -= totalCost;
                count += freq[cost];
            } else {
                count += coins / cost;
                break;
            }
        }

        return count;
    }
}
